### DISCUSSION TITLE
Complementary color trait for colors that have a `Hue`

### DISCUSSION THREAD

**Discussion by explosion-mental** [Ogeon/palette#383] [Ideas]
While [switching/porting more stuff to `palette`](https://codeberg.org/explosion-mental/wallust/commit/b023a478001a3ab66bafe77b087e6dd22dbeb44d), I've noticed It doesn't have a `Complementary` color feature. Not sure how useful or needed this is, but the crate does give you enough tools to implement it yourself.

I have a dead simple prototype, I'm sure yourself as the author could come up with something more sophisticated and/or find some issues with it.

https://codeberg.org/explosion-mental/wallust/src/commit/b023a478001a3ab66bafe77b087e6dd22dbeb44d/src/colors.rs#L236-L270

**Comment by Ogeon** [comment]
Thanks for bringing this up! It can be nice to have a dedicated trait for it, even if it can be accomplished with hue shifts. Especially if it's possible to implement for more types than those with explicit hue components. The meaning of the terminology needs to be nailed down, though. It looks like your implementation and the one in [the linked stack overflow question](https://stackoverflow.com/questions/9577590/formula-to-find-the-split-complementaries-of-a-color#12014465) are quite different, not to mention that you are dividing the hues into chunks.

**Reply by explosion-mental** [reply]
> It looks like your implementation and the one in [the linked stack overflow question](https://stackoverflow.com/questions/9577590/formula-to-find-the-split-complementaries-of-a-color#12014465) are quite different, not to mention that you are dividing the hues into chunks.

Well that was because [I was using `easy_colors`](https://codeberg.org/explosion-mental/wallust/src/commit/7c52fb414745d486a4e4fd9c9e0a5aadb8c42012/src/colors.rs#L246-L266) which didn't have anything like `shift_hue()`. Since shift_hue on the palette crate can understand and treats values within 0..=360 range (or some of it's siblings functions) a simple `shift_hue(180)` should suffice (and maybe even add [other counterparts](https://codeberg.org/explosion-mental/wallust/issues/13#issue-333910)).

I was thinking of my implementation as [this Newton's wheel](https://net-informations.com/q/web/wheel.html) (reminds me of the circle of fifths), so I wanted to preserve the distance that the color was positioned in the current arc, which `shift_huw` alrd does, since it adds/subtracts the amount to the hue (if i'm not mistaken).

As for the linked stack overflow, I'm not sure why they add/subtract 30, since in hsv it's defined ([Complementary Colors - Wikipedia](https://en.wikipedia.org/wiki/Complementary_colors#Colors_produced_by_light)):
> In some other color models, such as the [HSV color space](https://en.wikipedia.org/wiki/HSL_and_HSV), the neutral colors (white, grays, and black) lie along a central axis. Complementary colors (as defined in HSV) lie opposite each other on any horizontal cross-section. For example, in the [CIE 1931 color space](https://en.wikipedia.org/wiki/CIE_1931_color_space) a color of a "[dominant](https://en.wikipedia.org/wiki/Dominant_wavelength)" wavelength can be mixed with an amount of the [complementary wavelength](https://en.wikipedia.org/wiki/Complementary_wavelength) to produce a neutral color (gray or white).

**Reply by Ogeon** [reply]
I see, your use case may need something more hand crafted in the end. That's ok, and a trait for complementary colors is still valuable on its own.

I just wanted to be sure we are on the same page, and it seems like we both mean a 180 degree rotation. I.e. the Wikipedia definition. The stack overflow question is talking about "split complementary", which is two colors that are adjacent to the complementary. You will also see people talking about triadic, tetriadic and analogous colors. All of these could probably be traits for convenience.

**Reply by explosion-mental** [reply]
Sounds good. Would love to [help](https://github.com/Ogeon/palette/issues/372) but I think I'm not that knowledgeable on color theory. I'll be active thanks to the project I'm developing (and trying to reach a stable version) so don't mind my newbie questions :)

**Reply by Ogeon** [reply]
No worries! The basics are quite simple. I got around to implement the common color schemes in #384. The module documentation is still a bit sparse, but let me know if anything seems weird, in case you decide to have a look. :slightly_smiling_face:
